geo/src/algorithm/triangulate_earcut.rs

```diff
@@ -0,0 +1,259 @@
+    /// let square_polygon = polygon![
+    ///     (x: 0., y: 0.), // SW
+    ///     (x: 10., y: 0.), // SE
+    ///     (x: 10., y: 10.), // NE
+    ///     (x: 0., y: 10.), // NW
+    ///     (x: 0., y: 0.), // SW
+    /// ];
+    ///
+    /// let mut triangles_iter = square_polygon.triangulate_earcut_iter();
+    ///
+    /// assert_eq!(
+    ///     Some(Triangle(
+    ///             coord! { x: 0., y: 10. }, // NW
+    ///             coord! { x: 10., y: 10. }, // NE
+    ///             coord! { x: 10., y: 0. }, // SE
+    ///     )),
+    ///     triangles_iter.next(),
+    /// );
+    ///
+    /// assert_eq!(
+    ///     Some(Triangle(
+    ///         coord! { x: 10., y: 0. }, // SE
+    ///         coord! { x: 0., y: 0. }, // SW
+    ///         coord! { x: 0., y: 10. }, // NW
+    ///     )),
+    ///     triangles_iter.next(),
+    /// );
+    ///
+    /// assert!(triangles_iter.next().is_none());
+    /// ```
+    fn triangulate_earcut_iter(&self) -> Iter<T> {
+        Iter(self.triangulate_earcut_raw())
+    }
+
+    /// Return the raw result from the `earcutr` library: a one-dimensional vector of polygon
+    /// vertices (in XY order), and the indicies of the triangles within the vertices vector. This
+    /// method is less ergonomic than the `triangulate_earcut` and `triangulate_earcut_iter`
+    /// methods, but can be helpful when working in graphics contexts that expect flat vectors of
+    /// data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::{coord, polygon, Triangle, TriangulateEarcut};
+    /// use geo::triangulate_earcut::Raw;
+    ///
+    /// let square_polygon = polygon![
+    ///     (x: 0., y: 0.), // SW
+    ///     (x: 10., y: 0.), // SE
+    ///     (x: 10., y: 10.), // NE
+    ///     (x: 0., y: 10.), // NW
+    ///     (x: 0., y: 0.), // SW
+    /// ];
+    ///
+    /// let mut triangles_raw = square_polygon.triangulate_earcut_raw();
+    ///
+    /// assert_eq!(
+    ///     Raw {
+    ///         vertices: vec![
+    ///             0., 0., // SW
+    ///             10., 0., // SE
+    ///             10., 10., // NE
+    ///             0., 10., // NW
+    ///             0., 0., // SW
+    ///         ],
+    ///         triangle_indices: vec![
+    ///             3, 0, 1, // NW-SW-SE
+    ///             1, 2, 3, // SE-NE-NW
+    ///         ],
+    ///     },
+    ///     triangles_raw,
+    /// );
+    /// ```
+    fn triangulate_earcut_raw(&self) -> Raw<T>;
+}
+
+impl<T: CoordFloat> TriangulateEarcut<T> for Polygon<T> {
+    fn triangulate_earcut_raw(&self) -> Raw<T> {
+        let input = polygon_to_earcutr_input(self);
+        let triangle_indices =
+            earcutr::earcut(&input.vertexes, &input.interior_indexes, 2).unwrap();
+        Raw {
+            vertices: input.vertexes,
+            triangle_indices,
+        }
+    }
+}
+
+/// The raw result of triangulating a polygon from `earcutr`.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Raw<T: CoordFloat> {
```

Comment:
Since this is public, I feel like a more descriptive name is in order. Maybe `Triangulation` ?

Reply:
Perhaps worth bearing in mind that if we went with Triangulation it would also have to be suitable for any future triangulations (e.g. Delaunay) we might add.

Reply:
This is name spaced to the `triangulate_earcut::Triangulation` module (v.s. the hypothetical `delaunay::Triangulation`), so I think it'd be fine.

That said, I'm also fine with the clarity/verboseness tradeoff of `triangulate_earcut::EarcutTriangulation`

I just think `Raw` is a bit inexplicable.

Reply:
There is a case to be made for `triangulate_earcut::Triangulation`, but I wonder if we should have a larger discussion about naming and namespacing to be consistent with our other traits. For example we also don't do `geo::euclidean_distance::Distance` nor `geo::geodeisc_length::Length`.

Reply:
> geo::euclidean_distance::Distance nor geo::geodeisc_length::Length

I don't understand the comparison to EuclideanDistance and GeodesicLength which both return scalars.  We don't currently have bespoke types for those operations. 

We *could* introduce something like:
```diff
 mod euclidean_distance {
+    struct Distance<T: CoordNum>(T);
     trait EuclideanDistance<T> {
-         fn euclidean_distance(&self) -> T;
+         fn euclidean_distance(&self) -> Distance<T>;
     }
 }
```
Is that what you mean? I don't see the value in it, but maybe I'm missing something.

In the case of this triangulation module, we already *have* this bespoke type: `Raw`, I'm saying we should give it a more descriptive name than "Raw".

Reply:
Disregard my previous comment! I interpreted your message:

> This is namespaced to the module, so `triangulate_earcut::Triangulation` ...

to mean that you were suggesting renaming the `TriangulateEarcut` trait to `Triangulation` and forgot this was a discussion about the struct that is being returned.

Reply:
Ah, right! Sorry for not being clearer from the get-go.  I probably should have just led with:

```suggestion
pub struct Triangulation<T: CoordFloat> {
```